### ISSUE TITLE
How transitive is repr(transparent)?

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/unsafe-code-guidelines#486]
Context: https://github.com/hyperium/hyper/issues/3500

`repr(transparent)` makes types have the same representation/ABI. Does it have the same effect on generic types containing it?

For example, in the following case:

```rust
enum Generic<T> {
   Type(T),
   TwoType(T, T),
   String(String),
   Nothing
}

#[repr(transparent)]
struct Wrapper<T>(T);
```

is it safe to transmute between `Generic<T>` and `Generic<Wrapper<T>>`? Do they have the same ABI? I'm thinking about things like `-Zrandomize-layout`.

**Comment by RalfJung** [comment]
> Do they have the same ABI?

[No](https://doc.rust-lang.org/stable/std/primitive.fn.html#abi-compatibility).

> is it safe to transmute between Generic\<T> and Generic<Wrapper\<T>>? 

I think that s the intention; I don't know if it actually follows from anything we document.  I also don't know what `-Zrandomize-layout` does here.

**Comment by Manishearth** [comment]
Probably worth documenting somewhere.

I suspect `-Zrandomize-layout` doesn't respect this, but that's fine, it's not a perfect representation of the rules.

**Comment by RalfJung** [comment]
That would make randomize-layout unsound though, which is clearly a bug. So I don't think it's "fine".

**Comment by digama0** [comment]
I don't think we promise that `Generic<T>` and `Generic<Wrapper<T>>` are the same. In general we can't do this, because of associated types. Maybe we can carve out some subset of cases for which it is okay, but I don't think that `-Zrandomize-layout` is doing anything unsound currently.

**Comment by RalfJung** [comment]
The question was for this specific type, which doesn't use associated types.

**Comment by digama0** [comment]
We definitely don't promise this, and `randomize-layout` may be laying them out differently. I don't think this is particularly desirable, but repr(Rust) structs have basically no layout guarantees.

**Comment by Manishearth** [comment]
> That would make randomize-layout unsound though, which is clearly a bug. So I don't think it's "fine".

Oh sure I meant that I wouldn't be surprised and I wouldn't treat it as a string expectation of what the semantics should be.

**Comment by FeldrinH** [comment]
> > Do they have the same ABI?
> 
> [No](https://doc.rust-lang.org/nightly/std/primitive.fn.html#abi-compatibility).

I read through the linked document and could not find a place that stated that the answer is no. What am I missing?

**Comment by RalfJung** [comment]
The answer is always "no" except when the linked document states otherwise. So the fact that none of the cases there says they are compatible implies that they are not compatible.

**Comment by FeldrinH** [comment]
> The answer is always "no" except when the linked document states otherwise. So the fact that none of the cases there says they are compatible implies that they are not compatible.

The document does not seem to discuss ABI compatibility of structs at all. Does that mean that two different structs are never ABI compatible?

**Comment by RalfJung** [comment]
It mentions repr(transparent). But structs without that attribute are indeed not ABI compatible with anything else.
